internal/rbd/nodeserver.go

```diff
@@ -1221,31 +1230,100 @@ func (ns *NodeServer) xfsSupportsReflink() bool {
+	var (
+		imgInfo rbdImageMetadataStash
+		err     error
+	)
+	var metadataPath string
+	volumePath := req.GetStagingTargetPath()
+	if volumePath == "" {
+		if volumePath = req.GetVolumePath(); volumePath == "" {
+			return nil, status.Error(codes.InvalidArgument, "volume path must be provided")
+		}
+		pvName := extractPVNameFromPath(volumePath)
+		metadataPath = ns.StagingPath + pvName + "/globalmount"
+	} else {
+		metadataPath = volumePath
+		volumePath = volumePath + "/" + req.GetVolumeId()
+	}
 
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+	imgInfo, err = lookupRBDImageMetadataStash(metadataPath)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to find image metadata at stagingPath %q: %v", volumePath, err)
+
+		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	stat, err := os.Stat(targetPath)
+	return checkBlockDeviceStatus(ctx, volumePath, &imgInfo, ns.Mounter)
+}
+
+// checkBlockDeviceStatus checks the block device current status.
+func checkBlockDeviceStatus(
+	ctx context.Context,
+	volumePath string,
+	imgInfo *rbdImageMetadataStash,
+	mounter mount.Interface,
+) (*csi.NodeGetVolumeStatsResponse, error) {
+	var volCondition csi.VolumeCondition
+
+	stat, err := os.Stat(volumePath)
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "failed to get stat for targetpath %q: %v", targetPath, err)
+		volCondition.Abnormal = true
+		volCondition.Message = fmt.Sprintf("failed to get stat for volumepath %q: %v", volumePath, err)
+
+		return &csi.NodeGetVolumeStatsResponse{
+			VolumeCondition: &volCondition,
+		}, nil
+	}
+
+	// userspace mounters
+	var stdout, stderr string
+	if imgInfo.NbdAccess {
+		stdout, stderr, err = util.ExecCommand(ctx, "rbd-nbd", "list-mapped")
+		if err != nil || stderr != "" {
+			log.ErrorLog(ctx, "running rbd-nbd list-mapped failed with error:%v, stderr:%s", err, stderr)
+
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		if !strings.Contains(stdout, imgInfo.DevicePath) && !strings.Contains(stdout, imgInfo.ImageName) {
+			volCondition.Abnormal = true
+			volCondition.Message = fmt.Sprintf("rbd-nbd process not found for image: %s, device: %s",
+				imgInfo.String(), imgInfo.DevicePath)
+
+			return &csi.NodeGetVolumeStatsResponse{
+				VolumeCondition: &volCondition,
+			}, nil
+		}
 	}
 
+	var res *csi.NodeGetVolumeStatsResponse
 	if stat.Mode().IsDir() {
-		return csicommon.FilesystemNodeGetVolumeStats(ctx, ns.Mounter, targetPath)
+		res, err = csicommon.FilesystemNodeGetVolumeStats(ctx, mounter, volumePath)
+		res.VolumeCondition = &volCondition
+
+		return res, err
 	} else if (stat.Mode() & os.ModeDevice) == os.ModeDevice {
-		return blockNodeGetVolumeStats(ctx, targetPath)
+		res, err = blockNodeGetVolumeStats(ctx, volumePath)
+		res.VolumeCondition = &volCondition
+
+		return res, err
 	}
 
-	return nil, fmt.Errorf("targetpath %q is not a block device", targetPath)
+	return nil, fmt.Errorf("volumePath %q is not a block device", volumePath)
```

Comment:
errors that get returned from this function should be GRPC errors (`status.Error()`)